extensions/sql-database-projects/src/models/project.ts

```diff
@@ -45,23 +44,117 @@ export class Project {
 
 		// find all folders and files to include
 
-		for (const itemGroup of result['ItemGroup']) {
-			if (itemGroup['Build'] !== undefined) {
-				for (const fileEntry of itemGroup['Build']) {
-					this.files.push(this.createProjectEntry(fileEntry.$['Include'], EntryType.File));
-				}
+		for (let ig = 0; ig < this.projFileXmlDoc.documentElement.getElementsByTagName('ItemGroup').length; ig++) {
+			const itemGroup = this.projFileXmlDoc.documentElement.getElementsByTagName('ItemGroup')[ig];
+
+			for (let b = 0; b < itemGroup.getElementsByTagName('Build').length; b++) {
+				this.files.push(this.createProjectEntry(itemGroup.getElementsByTagName('Build')[b].getAttribute('Include'), EntryType.File));
 			}
 
-			if (itemGroup['Folder'] !== undefined) {
-				for (const folderEntry of itemGroup['Folder']) {
-					this.files.push(this.createProjectEntry(folderEntry.$['Include'], EntryType.Folder));
-				}
+			for (let f = 0; f < itemGroup.getElementsByTagName('Folder').length; f++) {
+				this.files.push(this.createProjectEntry(itemGroup.getElementsByTagName('Folder')[f].getAttribute('Include'), EntryType.Folder));
 			}
 		}
 	}
 
+	public async writeProjFile() {
+		try {
+			const projFileText = '';
+			await fs.writeFile(this.projectFilePath, projFileText);
+		}
+		catch (err) {
+			vscode.window.showErrorMessage(err);
+			return;
+		}
+	}
+
 	private createProjectEntry(relativePath: string, entryType: EntryType): ProjectEntry {
-		return new ProjectEntry(vscode.Uri.file(path.join(this.projectFile, relativePath)), entryType);
+		return new ProjectEntry(vscode.Uri.file(path.join(this.projectFolderPath, relativePath)), relativePath, entryType);
+	}
+
+	public async addFolderItem(relativeFolderPath: string): Promise<ProjectEntry> {
+		const absoluteFolderPath = path.join(this.projectFolderPath, relativeFolderPath);
+		await fs.mkdir(absoluteFolderPath, { recursive: true });
+
+		const folderEntry = this.createProjectEntry(relativeFolderPath, EntryType.Folder);
+		this.files.push(folderEntry);
+
+		await this.addToProjFile(folderEntry);
+		return folderEntry;
+	}
+
+	public async addScriptItem(relativeFilePath: string, contents: string): Promise<ProjectEntry> {
+		const absoluteFilePath = path.join(this.projectFolderPath, relativeFilePath);
+		await fs.mkdir(path.dirname(absoluteFilePath), { recursive: true });
+		await fs.writeFile(absoluteFilePath, contents);
+
+		const fileEntry = this.createProjectEntry(relativeFilePath, EntryType.File);
+		this.files.push(fileEntry);
+
+		await this.addToProjFile(fileEntry);
+
+		return fileEntry;
+	}
+
+	private findOrCreateItemGroup(containedTag?: string): any {
+		let outputItemGroup = undefined;
+
+		// find any ItemGroup node that contains files; that's where we'll add
```

Comment:
I don't really mind this but I am worried about this for complex files... can we try offloading some of it to SqlToolsService and use Project class... that's a pretty simple implementation. The calling functions to add file and folder will all be here but the actual addition happens through Project class... any thoughts?

Reply:
We've chatted a bit about this since you made this comment, but to capture it here: I like the idea of moving project file manipulation down to SqlToolsService, but you think that given we'll likely be able to repurpose/expand the existing services for build/deploy, a new Projects service is looking less necessary.  It's something we like for the future as need (re-)arises.